File: compatibility-tests/sdk-test-node/tests/gcs-preconditions.test.ts
import { describe, it, expect, beforeAll, afterAll } from 'vitest';
import { Storage } from '@google-cloud/storage';
import { ENDPOINT, PROJECT_ID, uniqueName } from './setup';

// Regression coverage for issue #2: preconditions were accepted but ignored, so
// unsafe writes silently succeeded. Real GCS returns 412 Precondition Failed.

// The status surfaces differently across SDK code paths (resumable upload vs.
// metadata patch), so probe the common fields.
function statusOf(err: unknown): number | undefined {
  const e = err as { code?: number; status?: number; response?: { status?: number } };
  return e.code ?? e.status ?? e.response?.status;
}

async function expectStatus(promise: Promise<unknown>, status: number): Promise<void> {
  await promise.then(
    () => {
      throw new Error(`expected rejection with status ${status}`);
    },
    (err) => {
      expect(statusOf(err)).toBe(status);
    },
  );
}

describe('GCS preconditions', () => {
  let storage: Storage;
  let bucketName: string;
  const objectName = 'o';
  let generation: number;

  beforeAll(async () => {
    storage = new Storage({ apiEndpoint: ENDPOINT, projectId: PROJECT_ID });
    bucketName = uniqueName('precond-bucket');
    await storage.createBucket(bucketName);
    await storage.bucket(bucketName).file(objectName).save('v1');
    const [metadata] = await storage.bucket(bucketName).file(objectName).getMetadata();
    generation = Number(metadata.generation);
  });

  afterAll(async () => {
    await storage.bucket(bucketName).file(objectName).delete().catch(() => {});
    await storage.bucket(bucketName).delete().catch(() => {});
  });

  it('ifGenerationMatch=0 on an existing object fails with 412', async () => {
    const file = storage.bucket(bucketName).file(objectName);
    await expectStatus(file.save('v2', { preconditionOpts: { ifGenerationMatch: 0 } }), 412);
  });

  it('a stale ifGenerationMatch fails with 412', async () => {
    const file = storage.bucket(bucketName).file(objectName);

    // A matching generation precondition succeeds and bumps the generation.
    await file.save('v3', { preconditionOpts: { ifGenerationMatch: generation } });

    await expectStatus(file.save('v4', { preconditionOpts: { ifGenerationMatch: generation } }), 412);
  });

  it('a wrong ifMetagenerationMatch fails with 412', async () => {
    const file = storage.bucket(bucketName).file(objectName);
    await expectStatus(
      file.setMetadata({ metadata: { a: 'b' } }, { ifMetagenerationMatch: 999 }),
      412,
    );
  });
});

File: compatibility-tests/sdk-test-java/src/test/java/io/floci/gcp/test/GcsPreconditionsTest.java
package io.floci.gcp.test;

import com.google.cloud.storage.Blob;
import com.google.cloud.storage.BlobId;
import com.google.cloud.storage.BlobInfo;
import com.google.cloud.storage.BucketInfo;
import com.google.cloud.storage.Storage;
import com.google.cloud.storage.StorageException;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.MethodOrderer;
import org.junit.jupiter.api.Order;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestMethodOrder;

import java.nio.charset.StandardCharsets;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

/**
 * Regression coverage for issue #2: preconditions ({@code ifGenerationMatch},
 * {@code ifMetagenerationMatch}, ...) were accepted but never evaluated, so
 * unsafe writes silently succeeded. Real GCS returns 412 Precondition Failed.
 */
@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
class GcsPreconditionsTest {

    private static final String BUCKET_NAME = TestFixtures.uniqueName("precond-bucket");
    private static final String OBJECT_NAME = "obj.txt";

    private static Storage storage;
    private static long currentGeneration;

    @BeforeAll
    static void setUp() {
        storage = TestFixtures.storageClient();
        storage.create(BucketInfo.of(BUCKET_NAME));
        Blob blob = storage.create(BlobInfo.newBuilder(BlobId.of(BUCKET_NAME, OBJECT_NAME)).build(),
                "v1".getBytes(StandardCharsets.UTF_8));
        currentGeneration = blob.getGeneration();
    }

    @AfterAll
    static void tearDown() throws Exception {
        if (storage instanceof AutoCloseable closeable) {
            closeable.close();
        }
    }

    @Test
    @Order(1)
    void doesNotExistOnExistingObjectFailsWith412() {
        BlobInfo info = BlobInfo.newBuilder(BlobId.of(BUCKET_NAME, OBJECT_NAME)).build();
        assertThatThrownBy(() -> storage.create(info, "v2".getBytes(StandardCharsets.UTF_8),
                Storage.BlobTargetOption.doesNotExist()))
                .isInstanceOf(StorageException.class)
                .satisfies(e -> assertThat(((StorageException) e).getCode()).isEqualTo(412));
    }

    @Test
    @Order(2)
    void staleGenerationMatchFailsWith412() {
        BlobInfo info = BlobInfo.newBuilder(BlobId.of(BUCKET_NAME, OBJECT_NAME)).build();

        // A matching generation precondition succeeds and bumps the generation.
        Blob updated = storage.create(info, "v3".getBytes(StandardCharsets.UTF_8),
                Storage.BlobTargetOption.generationMatch(currentGeneration));
        assertThat(updated.getGeneration()).isNotEqualTo(currentGeneration);

        // Re-using the now-stale generation must fail.
        assertThatThrownBy(() -> storage.create(info, "v4".getBytes(StandardCharsets.UTF_8),
                Storage.BlobTargetOption.generationMatch(currentGeneration)))
                .isInstanceOf(StorageException.class)
                .satisfies(e -> assertThat(((StorageException) e).getCode()).isEqualTo(412));

        currentGeneration = updated.getGeneration();
    }

    @Test
    @Order(3)
    void wrongMetagenerationMatchFailsWith412() {
        BlobInfo info = BlobInfo.newBuilder(BlobId.of(BUCKET_NAME, OBJECT_NAME))
                .setContentType("application/json")
                .build();
        assertThatThrownBy(() -> storage.update(info, Storage.BlobTargetOption.metagenerationMatch(999)))
                .isInstanceOf(StorageException.class)
                .satisfies(e -> assertThat(((StorageException) e).getCode()).isEqualTo(412));
    }

    @Test
    @Order(4)
    void matchingMetagenerationSucceeds() {
        Blob blob = storage.get(BlobId.of(BUCKET_NAME, OBJECT_NAME));
        BlobInfo info = BlobInfo.newBuilder(BlobId.of(BUCKET_NAME, OBJECT_NAME))
                .setContentType("text/markdown")
                .build();
        Blob updated = storage.update(info,
                Storage.BlobTargetOption.metagenerationMatch(blob.getMetageneration()));
        assertThat(updated.getContentType()).isEqualTo("text/markdown");
    }
}
